src/security/Handshake.cc

```diff
@@ -631,9 +641,12 @@ Security::HandshakeParser::parseSupportedVersionsExtension(const SBuf &extension
 void
 Security::HandshakeParser::skipMessage(const char *description)
 {
-    // tkMessages/fragments can only contain messages of the same ContentType.
+    // skip at least one byte
+    if (tkMessages.atEnd())
+        throw Parser::InsufficientInput();
```

Comment:
I tried to implement this approach (up to 47eb627).

> The above XXX

I added this XXX for now but I am not sure how 'empty message sequences' are possible here.
HandshakeParser::parseModernRecord() guarantees that tkMessages will get at least one byte here (due to the protection 'MUST NOT send zero-length [non-application] fragments') - i.e., exhausted() should be false. Also, 'done' should have been checked earlier in parseHello().

Reply:
> I added this XXX for now but I am not sure how 'empty message sequences' are possible here. HandshakeParser::parseModernRecord() guarantees that tkMessages will get at least one byte here (due to the protection 'MUST NOT send zero-length [non-application] fragments')

Do you mean that we are "protected" by a "MUST NOT" text in an RFC? Surely an attacker can violate that MUST -- we cannot rely on that kind of protection or guarantees in a parser... AFAICT, the sketched code will just ignore/skip such invalid messages (i.e. there are no infinite loops or similar dangers here). This is not the end of the world, but to the parseNonEmptyMessages() caller it would look like a non-empty message was successfully parsed and that feels wrong, especially when some future refactoring might unknowingly create an infinite loop when handling such an invalid input (because all the method name suggests that this code should not worry about empty messages).

> Also, 'done' should have been checked earlier in parseHello().

Yes, but that messageParser() call inside the loop can make `done` truthy, so we have to check on every iteration AFAICT.

Reply:
> Do you mean that we are "protected" by a "MUST NOT" text in an RFC? 

No, I meant the protection in HandshakeParser::parseModernRecord():
```
Must(record.fragment.length() || record.type == ContentType::ctApplicationData);
```
With this protection, only ctApplicationData can be empty, which are handled by skipPossiblyEmptyMessages().

Reply:
> No, I meant [Must(record.fragment.length()...)] in HandshakeParser::parseModernRecord

Ah, I see. I agree that the above Must() protects parseNonEmptyMessages() from empty input. 

However, it is difficult to _assert_ that protection using the current code structure. I think we should use that protection to avoid the XXX without introducing assertions. We are expecting at least one byte, so we should parse input _before_ checking whether it is exhausted:

```C++
tkMessages.rollback();
do {
    parseMessage();
    tkMessages.commit();
} while (!done && !tkMessages.exhausted());
```

Reply:
Ok, adjusted at 4b525e4.